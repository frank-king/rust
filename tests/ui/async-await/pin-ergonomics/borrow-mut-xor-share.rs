#![feature(pin_ergonomics)]
#![allow(dead_code, incomplete_features)]

// Makes sure `&pin mut place` and `&pin const place` cannot violate the mut-xor-share rules.

use std::pin::Pin;

struct Foo;

fn foo_mut(_: &mut Foo) {
}

fn foo_ref(_: &Foo) {
}

fn foo_pin_mut(_: Pin<&mut Foo>) {
}

fn foo_pin_ref(_: Pin<&Foo>) {
}

fn bar() {
    let foo = Foo;
    foo_pin_mut(&pin mut foo); //~ ERROR cannot borrow `foo` as mutable, as it is not declared as mutable

    let mut foo = Foo;
    let x = &pin mut foo;
    foo_pin_ref(&pin const foo); //~ ERROR cannot borrow `foo` as immutable because it is also borrowed as mutable
    foo_pin_mut(&pin mut foo); //~ ERROR cannot borrow `foo` as mutable more than once at a time
    foo_ref(&foo); //~ ERROR cannot borrow `foo` as immutable because it is also borrowed as mutable
    foo_mut(&mut foo); //~ ERROR cannot borrow `foo` as mutable more than once at a time

    foo_pin_mut(x);

    let mut foo = Foo;
    let x = &pin const foo;
    foo_pin_ref(&pin const foo); // ok
    foo_pin_mut(&pin mut foo); //~ ERROR cannot borrow `foo` as mutable because it is also borrowed as immutable
    foo_ref(&foo); // ok
    foo_mut(&mut foo); //~ ERROR cannot borrow `foo` as mutable because it is also borrowed as immutable

    foo_pin_ref(x);

    let mut foo = Foo;
    let x = &mut foo;
    foo_pin_ref(&pin const foo); //~ ERROR cannot borrow `foo` as immutable because it is also borrowed as mutable
    foo_pin_mut(&pin mut foo); //~ ERROR cannot borrow `foo` as mutable more than once at a time

    foo_mut(x);

    let mut foo = Foo;
    let x = &foo;
    foo_pin_ref(&pin const foo); // ok
    foo_pin_mut(&pin mut foo); //~ ERROR cannot borrow `foo` as mutable because it is also borrowed as immutable

    foo_ref(x);
}

fn main() {}
