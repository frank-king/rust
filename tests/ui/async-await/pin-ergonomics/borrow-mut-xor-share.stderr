error[E0596]: cannot borrow `foo` as mutable, as it is not declared as mutable
  --> $DIR/borrow-mut-xor-share.rs:24:17
   |
LL |     foo_pin_mut(&pin mut foo);
   |                 ^^^^^^^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
LL |     let mut foo = Foo;
   |         +++

error[E0502]: cannot borrow `foo` as immutable because it is also borrowed as mutable
  --> $DIR/borrow-mut-xor-share.rs:28:17
   |
LL |     let x = &pin mut foo;
   |             ------------ mutable borrow occurs here
LL |     foo_pin_ref(&pin const foo);
   |                 ^^^^^^^^^^^^^^ immutable borrow occurs here
...
LL |     foo_pin_mut(x);
   |                 - mutable borrow later used here

error[E0499]: cannot borrow `foo` as mutable more than once at a time
  --> $DIR/borrow-mut-xor-share.rs:29:17
   |
LL |     let x = &pin mut foo;
   |             ------------ first mutable borrow occurs here
LL |     foo_pin_ref(&pin const foo);
LL |     foo_pin_mut(&pin mut foo);
   |                 ^^^^^^^^^^^^ second mutable borrow occurs here
...
LL |     foo_pin_mut(x);
   |                 - first borrow later used here

error[E0502]: cannot borrow `foo` as immutable because it is also borrowed as mutable
  --> $DIR/borrow-mut-xor-share.rs:30:13
   |
LL |     let x = &pin mut foo;
   |             ------------ mutable borrow occurs here
...
LL |     foo_ref(&foo);
   |             ^^^^ immutable borrow occurs here
...
LL |     foo_pin_mut(x);
   |                 - mutable borrow later used here

error[E0499]: cannot borrow `foo` as mutable more than once at a time
  --> $DIR/borrow-mut-xor-share.rs:31:13
   |
LL |     let x = &pin mut foo;
   |             ------------ first mutable borrow occurs here
...
LL |     foo_mut(&mut foo);
   |             ^^^^^^^^ second mutable borrow occurs here
LL |
LL |     foo_pin_mut(x);
   |                 - first borrow later used here

error[E0502]: cannot borrow `foo` as mutable because it is also borrowed as immutable
  --> $DIR/borrow-mut-xor-share.rs:38:17
   |
LL |     let x = &pin const foo;
   |             -------------- immutable borrow occurs here
LL |     foo_pin_ref(&pin const foo); // ok
LL |     foo_pin_mut(&pin mut foo);
   |                 ^^^^^^^^^^^^ mutable borrow occurs here
...
LL |     foo_pin_ref(x);
   |                 - immutable borrow later used here

error[E0502]: cannot borrow `foo` as mutable because it is also borrowed as immutable
  --> $DIR/borrow-mut-xor-share.rs:40:13
   |
LL |     let x = &pin const foo;
   |             -------------- immutable borrow occurs here
...
LL |     foo_mut(&mut foo);
   |             ^^^^^^^^ mutable borrow occurs here
LL |
LL |     foo_pin_ref(x);
   |                 - immutable borrow later used here

error[E0502]: cannot borrow `foo` as immutable because it is also borrowed as mutable
  --> $DIR/borrow-mut-xor-share.rs:46:17
   |
LL |     let x = &mut foo;
   |             -------- mutable borrow occurs here
LL |     foo_pin_ref(&pin const foo);
   |                 ^^^^^^^^^^^^^^ immutable borrow occurs here
...
LL |     foo_mut(x);
   |             - mutable borrow later used here

error[E0499]: cannot borrow `foo` as mutable more than once at a time
  --> $DIR/borrow-mut-xor-share.rs:47:17
   |
LL |     let x = &mut foo;
   |             -------- first mutable borrow occurs here
LL |     foo_pin_ref(&pin const foo);
LL |     foo_pin_mut(&pin mut foo);
   |                 ^^^^^^^^^^^^ second mutable borrow occurs here
LL |
LL |     foo_mut(x);
   |             - first borrow later used here

error[E0502]: cannot borrow `foo` as mutable because it is also borrowed as immutable
  --> $DIR/borrow-mut-xor-share.rs:54:17
   |
LL |     let x = &foo;
   |             ---- immutable borrow occurs here
LL |     foo_pin_ref(&pin const foo); // ok
LL |     foo_pin_mut(&pin mut foo);
   |                 ^^^^^^^^^^^^ mutable borrow occurs here
LL |
LL |     foo_ref(x);
   |             - immutable borrow later used here

error: aborting due to 10 previous errors

Some errors have detailed explanations: E0499, E0502, E0596.
For more information about an error, try `rustc --explain E0499`.
